Flexible table

araujoarthur0 opened the pull request:
#### Related issue
Closes #357

#### Context / Background
This PR is adding a flexible number of columns to TTL in the format of a new preference.
The user will be able to add/remove pairs of entries from the default "4" entries, and still keep balance and total information.

#### What change is being introduced by this PR?
I added a new class representing a Month Flexible Calendar, with a new Db store allowing for the flexible format. We're now saving times in array format in the Db, and I updated the clear/import/export db methods to include the new db as well.

One is able to control the #entries by clicking on the new +/- icons on the sides of the table rows. The new arrow icons allow horizontal movement, and so does scrolling with the mouse (a js dependency allows horizontal scroll without shift key).
![flex](https://user-images.githubusercontent.com/37311270/91928960-bc428a80-ecb3-11ea-8be1-0989546caa38.gif)

The table is resizable in a way that makes sense.
![flex2](https://user-images.githubusercontent.com/37311270/91929013-e431ee00-ecb3-11ea-9c28-b8f1904b4c90.gif)

And when you first move to the new flex calendar, you'll get the request to migrate your existing data. There's also an option on the "Edit" menu later if you miss the request.
![flex3](https://user-images.githubusercontent.com/37311270/91929040-ef851980-ecb3-11ea-8ec4-61500e2168c3.gif)

#### How will this be tested?
Still going to add tests just for the new Calendar, but all the rest are ok.

araujoarthur0 commented:
I wanted to create the pull request at once instead of waiting for a Day calendar flex version, since that might take even more time with a stale PR.
I might open another issue for that and for the tests.

thamara commented:
From the animations, the feature looks awesome! 
I'm review it carefully at the weekend and let you know :)

araujoarthur0 commented:
Added some tests, split files and formatted others.

araujoarthur0 commented:
macos again :( and I have no idea why it's complaining about pure javascript now on the Calendar test, from which I just removed things

thamara commented:
The failing test was just a syntax error. I have already fixed that or you :)

araujoarthur0 commented:
> The failing test was just a syntax error. I have already fixed that or you :)

Nice catch! I have no idea how I introduced this change in https://github.com/thamara/time-to-leave/pull/358/commits/0831bd3452353966978ae0ce81fd10365fddc433, but I'm reverting it too cause it's nonsense heuaheua

araujoarthur0 commented on js/classes/FlexibleMonthCalendar.js:
```diff
@@ -0,0 +1,702 @@
+    static _getRowCode(dateKey, isInterval = false) {
+        if (isInterval) {
+            return  '<div class="row-time">' +
+                        '<div class="interval" colspan="4"><span>--:--</span></div>' +
```
Changed it to an empty space like on default calendar :)
I tried avoiding the input since it won't really be un-disabled, so it's a span containing the values.

araujoarthur0 commented:
Cleaned CodeFactor issues aside from ComplexCode ones

thamara commented on __tests__/__main__/time-balance.js:
```diff
@@ -282,4 +240,211 @@ describe('Time Balance', () => {
+
+    test('getFirstInputInDb: input 4', () => {
+        const entryEx = {
+            '2020-6-6': {'values': ['10:00', '12:00', '13:00', '14:00']},
+            '2020-6-7': {'values': ['10:00', '12:00', '13:00', '14:00']},
+            '2020-6-8': {'values': ['10:00', '13:00', '14:00', '19:00']},
+            '2020-6-9': {'values': ['10:00', '12:00', '13:00', '22:00']},
+            '2020-6-10': {'values': ['08:00', '12:00', '13:00', '19:00']}
```
In all of these tests, you are considering 4 entries, which would be equivalent to the same behavior we have today. Maybe in the future, we could refactor the tests to be able to test both in a more seamless way.
But what I think it's important to include now, are some tests with more than 4 entries, and also fewer entries, which would also be possible now, right?

araujoarthur0 replied:
Added some irregular tests too, nice catch :)

thamara commented on changelog.txt:
```diff
@@ -4,6 +4,7 @@
 - Enhancement: [#333] Adding start date for overall balance on preferences
+- Enhancement: [#357] Adding flexible table format for month calendar with variable number of entries per day
 
 Who built 1.5.6:
```
Add your pretty name here too :)

thamara commented on js/classes/FlexibleMonthCalendar.js:
```diff
@@ -0,0 +1,709 @@
+            toggleArrowColor(element);
+        });
+
+        function removeEntries(element) {
```
I think before removing row times that are not empty, we should issue a confirmation pop up. I think it might be fairly easy for someone to incorrectly press the - button leading the person to lose the data.

araujoarthur0 replied:
Added the popup, and also changed the button to be disabled when there's nothing to remove.

thamara commented on js/classes/FlexibleMonthCalendar.js:
```diff
@@ -0,0 +1,709 @@
+    /*
+    * Adds the next missing entry on the actual day and updates calendar.
+    */
+    punchDate() {
```
I see a few enhancements for the future on this:
- allow to punch time, without needing to allow more rows (TTL would automatically do that)
- When allowing such behavior, have some kind of focus on the newly added entry (for example, scrolling to the line

For now, what I noticed is that if I click on the plus sign, adding more columns to the day, I still cannot use the punch time more than 4 times.

araujoarthur0 replied:
I extracted a function to review the punch button for the day and it now does these expected checks.

thamara commented on js/classes/FlexibleMonthCalendar.js:
```diff
@@ -0,0 +1,709 @@
+        $('.plus-sign span').off('click').on('click', function() {
+            const element = $(this).parent().parent().find('.time-cells')[0];
+            addEntries(element);
+            toggleArrowColor(element);
```
We are missing a togglePunchButton here (if it was already disabled).

thamara commented on js/classes/FlexibleMonthCalendar.js:
```diff
@@ -0,0 +1,709 @@
+
+            i++;
+        }
+        while (i < 5 || i % 3 !== 2) {
+            if ((i + 1) % 3 === 2 || (i + 1) % 3 === 1) {
```
I think this math on the column number should be extracted to a function in order to improve the code redability.

thamara commented on js/import-export.js:
```diff
@@ -110,8 +153,38 @@ function importDatabaseFromFile(filename) {
+            regularEntryArray[key].values.sort();
+            flexibleStore.set(key, regularEntryArray[key]);
+        }
+    } catch (err) {
+        return false;
```
I think it's important to inform the user on a failure, possibly retrieving the error information so that they can send this to us.

araujoarthur0 replied:
Is a console.log enough? heuahea

thamara replied:
I think so. It should be enough for us to ask someone that got the problem to open the console and check the error.

thamara commented on main.js:
```diff
@@ -229,6 +268,21 @@ function createWindow() {
                 },
                 {type: 'separator'},
+                {
+                    label: 'Migrate to flexible calendar',
```
After migrating to flexible calendar, I think the option in the menu should be disabled.

araujoarthur0 commented:
@thamara, tried to force the EOF to be LF for .js files using https://github.com/thamara/time-to-leave/pull/358/commits/1adedcb70d0dde00642f577f3d3ecf1bfa8e2335, but no deal :( 
I'm confused, cause we haven't changed the lint rules, have we?

And the same is happening on your commits from Mac (https://github.com/thamara/time-to-leave/pull/364), so it may not be a windows thing.

thamara commented:
Yeah. I saw that too. I don't think we changed anything, but I'll investigate. For now I think you can ignore such errors.

araujoarthur0 commented:
Alright, I took off the EOL change. I'll be merging then :)
